operator/pkg/ciliumendpointbatch/aggregator.go

```diff
@@ -0,0 +1,272 @@
+	// in next CEBDelete reconcile sync.
+	// CebLazyUpdate is used only if CEPs are removed from CEB. For other CEP changes i.e.
+	// CEP Insertion or CEP modification in a CEB, CebUpdate is event is used instead.
+	CebLazyUpdate CebEventType = "CebsLazyUpdate"
+	// CebDelete is an event. it tells aggregator, to delete a CEB with k8s-apiserver
+	// in next CEBDelete reconcile sync.
+	CebDelete CebEventType = "CebsToDelete"
+)
+
+// aggregator is used to aggregate CEB events and resolve them into a single actionable item for the CEB.
+// CEB manager watches for CEP updates and enqueue or dequeue them in different CEB's and calls updateAggregator function.
+// aggregator job is to track what changes happened to a CEB and converge into a single update for reconcile purpose.
+type aggregator struct {
+	// syncMutex is used to protect data in aggregator and concurrent thread safe access.
+	syncMutex lock.RWMutex
+	// cebsCreate is a set of CEB names, contains list of CEB's to be created in next
+	// CEBUpsertSync reconcile period.
+	cebsCreate map[string]struct{}
+	// cebsUpdate is a set of CEB names, contains list of CEB's to be updated in next
+	// CEBUpsertSync reconcile period.
+	// This cebsUpdate set may contains following CEB's. If,
+	// 1) New CEP is inserted into the CEB or
+	// 2) Existing CEP values are modified in the CEB.
+	// 3) CEPs are removed from the CEB and either one of above case is true.
+	cebsUpdate map[string]struct{}
+	// cebsLazyUpdate is a set of CEB names, contains list of CEB's to be updated in
+	// next CEBDeleteSync reconcile period.
+	// This CEB lazyUpdate set contains following CEB's. If,
+	// 1) Only CEPs are removed from this CEB.
+	cebsLazyUpdate map[string]struct{}
+	// cebsDelete is a set, contains list of CEB's to be deleted in next CEBDeleteSync reconcile period.
+	cebsDelete map[string]struct{}
+}
+
+// Create and Initialize a aggregator
+func newAggregator() *aggregator {
+
+	return &aggregator{
+		cebsCreate:     make(map[string]struct{}),
+		cebsUpdate:     make(map[string]struct{}),
+		cebsLazyUpdate: make(map[string]struct{}),
+		cebsDelete:     make(map[string]struct{}),
+	}
+
+}
+
+// updateAggregator receives CEBName and event for various CEB's, it would converge
+// those requests into a single actionable item for a CEB.
+//
+// There are 4 different states, Create, Update, LazyUpdate and Delete and there are
+// four events CebDelete, CebCreate, CebUpdate and CebLazyUpdate.
+//
+// Based on the given event, CEB is placed in one of the state. Below is state machine for Delete state,
+//
+// Input-Event         Current-State     Output-State
+//  CebDelete               NA               Delete
+//  CebDelete             Delete             Delete
+//  CebCreate             Delete             Delete
+//  CebUpdate             Delete             Delete
+//  CebLazyUpdate         Delete             Delete
+// Delete is special state, its terminal state. Once CEB is marked for delete, then client
```

Comment:
Per our discussion offline, this is not true, if a previous CEB delete did not go through, and a re-created CEP was observed, this CEB would go to CebUpdate or CebCreate queue, no?

Reply:
I didn't take that route, if you look carefully in code. CEB is marked for DELETE only if all CEP's are deleted in a CEB and successfully synced with api-server. Until then CEB is in `CebUpdate` state. 
Moving the CEB from `CebDelete` to `CebCreate` have a race condition Issue and we need to clean up existing CEB `ObjectMeta` headers.

Reply:
I don't see the "only if all CEP's are deleted in a CEB and successfully synced with api-server." part, see my comment below.

Reply:
Below code wasn't removed correctly by mistake, updated now.